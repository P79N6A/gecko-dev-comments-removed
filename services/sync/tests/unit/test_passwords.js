Cu.import("resource://weave/util.js");
Cu.import("resource://weave/async.js");
Cu.import("resource://weave/dav.js");
Cu.import("resource://weave/identity.js");

let __fakePasswords = {
  'Mozilla Services Password': {foo: "bar"},
  'Mozilla Services Encryption Passphrase': {foo: "passphrase"}
};

let __fakePrefs = {
  "encryption" : "none",
  "log.logger.service.crypto" : "Debug",
  "log.logger.service.engine" : "Debug",
  "log.logger.async" : "Debug"
};

function run_test() {
  var fpasses = new FakePasswordService(__fakePasswords);
  var fprefs = new FakePrefService(__fakePrefs);
  var fds = new FakeDAVService({});
  var fts = new FakeTimerService();
  var logStats = initTestLogging();

  ID.set('Engine:PBE:default',
         new Identity('Mozilla Services Encryption Passphrase', 'foo'));

  
  var passwords = loadInSandbox("resource://weave/engines/passwords.js");

  
  var fakeUser = {
    hostname: "www.boogle.com",
    formSubmitURL: "http://www.boogle.com/search",
    httpRealm: "",
    username: "",
    password: "",
    usernameField: "test_person",
    passwordField: "test_password"
    };

  Utils.getLoginManager = function fake_getLoginManager() {
    
    return {getAllLogins: function() { return [fakeUser]; }};
  };

  Utils.getProfileFile = function fake_getProfileFile(arg) {
    return {exists: function() {return false;}};
  };

  Utils.open = function fake_open(file, mode) {
    let fakeStream = {
      writeString: function(data) {Log4Moz.Service.rootLogger.debug(data);},
      close: function() {}
    };
    return [fakeStream];
  };

  
  var fakeUserHash = passwords._hashLoginInfo(fakeUser);
  do_check_eq(typeof fakeUserHash, 'string');
  do_check_eq(fakeUserHash.length, 40);

  
  var psc = new passwords.PasswordSyncCore();
  do_check_false(psc._itemExists("invalid guid"));
  do_check_true(psc._itemExists(fakeUserHash));

  
  var engine = new passwords.PasswordEngine();
  let calledBack = false;

  function cb() {
    calledBack = true;
  }

  engine.sync(cb);

  while (fts.processCallback()) {}

  do_check_true(calledBack);
  do_check_eq(logStats.errorsLogged, 0);
  do_check_eq(Async.outstandingGenerators, 0);
}
