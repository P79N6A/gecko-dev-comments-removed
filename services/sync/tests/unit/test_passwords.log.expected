*** test pending
Service.PasswordEngine	INFO	Beginning sync
Testing	INFO	Creating dir user-data/passwords/
Service.RemoteStore	DEBUG	Downloading status file
Testing	INFO	Retrieving user-data/passwords/status.json, returning status 404
Service.PasswordEngine	INFO	Initial upload to server
Service.JsonFilter	DEBUG	Encoding data as JSON
Testing	INFO	Putting user-data/passwords/keys.json with data: {"ring":{}}
Service.Resource	DEBUG	PUT request successful
Service.JsonFilter	DEBUG	Encoding data as JSON
Service.CryptoFilter	DEBUG	Encrypting data
Testing	INFO	Putting user-data/passwords/snapshot.json with data: {"805ec58eb8dcded602999967e139be21acd0f194":{"hostname":"www.boogle.com","formSubmitURL":"http://www.boogle.com/search","httpRealm":"","username":"","password":"","usernameField":"test_person","passwordField":"test_password"}}
Service.Resource	DEBUG	PUT request successful
Service.JsonFilter	DEBUG	Encoding data as JSON
Service.CryptoFilter	DEBUG	Encrypting data
Testing	INFO	Putting user-data/passwords/deltas.json with data: []
Service.Resource	DEBUG	PUT request successful
Service.JsonFilter	DEBUG	Encoding data as JSON
Testing	INFO	Putting user-data/passwords/status.json with data: {"GUID":"fake-guid","formatVersion":2,"snapVersion":0,"maxVersion":0,"snapEncryption":"none","deltasEncryption":"none","itemCount":1}
Service.Resource	DEBUG	PUT request successful
Service.RemoteStore	INFO	Full upload to server successful
Service.SnapStore	INFO	Saving snapshot to disk
root	DEBUG	{"version":0,"GUID":"fake-guid","snapshot":{"805ec58eb8dcded602999967e139be21acd0f194":{"hostname":"www.boogle.com","formSubmitURL":"http://www.boogle.com/search","httpRealm":"","username":"","password":"","usernameField":"test_person","passwordField":"test_password"}}}
Testing	INFO	Initial sync done, re-syncing now.
Service.PasswordEngine	INFO	Beginning sync
Testing	INFO	Creating dir user-data/passwords/
Service.RemoteStore	DEBUG	Downloading status file
Testing	INFO	Retrieving user-data/passwords/status.json, returning status 200
Service.Resource	DEBUG	GET request successful
Service.JsonFilter	DEBUG	Decoding JSON data
Service.RemoteStore	DEBUG	Downloading status file... done
Service.PasswordEngine	INFO	Local snapshot version: 0
Service.PasswordEngine	INFO	Server maxVersion: 0
Service.RemoteStore	DEBUG	Using last sync snapshot as server snapshot (snap version == max version)
Service.RemoteStore	TRACE	Local snapshot version == server maxVersion
Service.PasswordEngine	INFO	Sync complete: no changes needed on client or server
Testing	INFO	Re-sync done, adding a login and re-syncing.
Service.PasswordEngine	INFO	Beginning sync
Testing	INFO	Creating dir user-data/passwords/
Service.RemoteStore	DEBUG	Downloading status file
Testing	INFO	Retrieving user-data/passwords/status.json, returning status 200
Service.Resource	DEBUG	GET request successful
Service.JsonFilter	DEBUG	Decoding JSON data
Service.RemoteStore	DEBUG	Downloading status file... done
Service.PasswordEngine	INFO	Local snapshot version: 0
Service.PasswordEngine	INFO	Server maxVersion: 0
Service.RemoteStore	DEBUG	Using last sync snapshot as server snapshot (snap version == max version)
Service.RemoteStore	TRACE	Local snapshot version == server maxVersion
Service.PasswordEngine	INFO	Reconciling client/server updates
Service.PasswordSync	DEBUG	Reconciling 1 against 0 commands
Service.PasswordEngine	INFO	Changes for client: 0
Service.PasswordEngine	INFO	Predicted changes for server: 1
Service.PasswordEngine	INFO	Client conflicts: 0
Service.PasswordEngine	INFO	Server conflicts: 0
Service.PasswordEngine	INFO	Actual changes for server: 1
Service.PasswordEngine	DEBUG	Actual changes for server: [{"action":"create","GUID":"1b3869fc36234b39cd354f661ed1d7d148394ca3","depth":0,"parents":[],"data":{"hostname":"www.yoogle.com","formSubmitURL":"http://www.yoogle.com/search","httpRealm":"","username":"","password":"","usernameField":"test_person2","passwordField":"test_password2"}}]
Service.PasswordEngine	INFO	Uploading changes to server
Testing	INFO	Retrieving user-data/passwords/deltas.json, returning status 200
Service.Resource	DEBUG	GET request successful
Service.CryptoFilter	DEBUG	Decrypting data
Service.JsonFilter	DEBUG	Decoding JSON data
Service.JsonFilter	DEBUG	Encoding data as JSON
Service.CryptoFilter	DEBUG	Encrypting data
Testing	INFO	Putting user-data/passwords/deltas.json with data: [[{"action":"create","GUID":"1b3869fc36234b39cd354f661ed1d7d148394ca3","depth":0,"parents":[],"data":{"hostname":"www.yoogle.com","formSubmitURL":"http://www.yoogle.com/search","httpRealm":"","username":"","password":"","usernameField":"test_person2","passwordField":"test_password2"}}]]
Service.Resource	DEBUG	PUT request successful
Service.JsonFilter	DEBUG	Encoding data as JSON
Testing	INFO	Putting user-data/passwords/status.json with data: {"GUID":"fake-guid","formatVersion":2,"snapVersion":0,"maxVersion":1,"snapEncryption":"none","deltasEncryption":"none","itemCount":2}
Service.Resource	DEBUG	PUT request successful
Service.PasswordEngine	INFO	Successfully updated deltas and status on server
Service.SnapStore	INFO	Saving snapshot to disk
root	DEBUG	{"version":1,"GUID":"fake-guid","snapshot":{"805ec58eb8dcded602999967e139be21acd0f194":{"hostname":"www.boogle.com","formSubmitURL":"http://www.boogle.com/search","httpRealm":"","username":"","password":"","usernameField":"test_person","passwordField":"test_password"},"1b3869fc36234b39cd354f661ed1d7d148394ca3":{"hostname":"www.yoogle.com","formSubmitURL":"http://www.yoogle.com/search","httpRealm":"","username":"","password":"","usernameField":"test_person2","passwordField":"test_password2"}}}
Service.PasswordEngine	INFO	Sync complete
*** test finished
*** exiting
*** PASS ***
